feat: Support numeric enums: BUILD file updates, new Go BUILD files

vchudnov-g opened the pull request:
This commit contains two changes:

1. When updating BUILD files, any pre-existing values of the
   `rest_numeric_enums` parameter will be preserved. This feature will
   be enabled for each language as they add support for that parameter
   in `BUILD.bazel.gapic_api.mustache`, which they'll typically do in
   separate CLs once they implement the parameter in their BUILD rule
   definition.

2. We also add support for `rest_numeric_enums` in
   `BUILD.bazel.gapic_api.mustache` for Go, since that will be the first
   language where the feature will be rolled out.

As much as we would have preferred to have these be separate commits,
they are entangled by the fact that testing (1) requires changes to
the template for one language, which is why we are doing (2) at the
same time.

vchudnov-g committed:
feat: Support numeric enums: BUILD file updates, new Go BUILD files

This commit contains two changes:

1. When updating BUILD files, any pre-existing values of the
   `rest_numeric_enums` parameter will be preserved. This feature will
   be enabled for each language as they add support for that parameter
   in BUILD.bazel.gapic_api.mustache, which they'lltypically do in
   separate CLs once they implement the parameter in their BUILD rule
   definition.

2. We also add support for `rest_numeric_enums` in
   BUILD.bazel.gapic_api.mustache for Go, since that will be the first
   language where the feature will be rolled out.

As much as we would have preferred to have these be separate commits,
they are entangled by the fact that testing (1) requires changes to
the template for one language, which is why we are doing (2) at the
same time.

alexander-fenster commented on bazel/src/main/java/com/google/api/codegen/bazel/ApiVersionedDir.java:
```diff
@@ -178,10 +173,28 @@ class ApiVersionedDir {
+    // Multiple languages:
+    PRESERVED_PROTO_LIBRARY_STRING_ATTRIBUTES.put("package_name", null);
+    PRESERVED_PROTO_LIBRARY_STRING_ATTRIBUTES.put("transport", null);
+    PRESERVED_PROTO_LIBRARY_STRING_ATTRIBUTES.put("rest_numeric_enums", "False");
```
would it be a string `"False"`, or a boolean `False`? I feel there is a possibility of an inconsistency here.

alexander-fenster replied:
Or, if I'm guessing correctly, for a string, it should be `"\"False\""`?

alexander-fenster replied:
After looking at the remaining code, I must say it is still unclear to me. Would it be possible to inspect the actual content of the build file that had `False` value added? Something tells me that it would be a string `"False"` rather than a boolean.

vchudnov-g replied:
This is intended to be a boolean value as per the internal design doc, which is reflected in the [Go](https://github.com/googleapis/gapic-generator-go/blob/4f333a3e023bdc6e384a8bc95409ba48f7b04727/rules_go_gapic/go_gapic.bzl#L96) and [Python](https://github.com/googleapis/gapic-generator-python/blob/8839c6fc593d62a10a069cb28f72cd76f080ecc5/rules_python_gapic/py_gapic.bzl#L46) implementations. For sample files, you can look at the mustache and baseline files in this PR. Or were you looking for something else?

noahdietz replied:
Definitely a boolean `False`.

codyoss reviewed (commented):
LGTM

noahdietz reviewed (approved):
LGTM from Go side

vam-google reviewed (approved):
LGTM

vchudnov-g committed:
Create a baseline file for BUILD file updates

This helps ensure that updated parameters are quoted or not, as expected.

Note that with this commit, updated values of rest_numeric_enums are quoted, which is incorrect.
TODO next: Fix this

vchudnov-g committed:
Separate overriden string and non-string attributes

vchudnov-g committed:
Write non-string overridden attributes unquoted

vchudnov-g committed:
Update tests to use overrides different than template values

vchudnov-g committed:
Use entrySet() when traversing non-string overrides.

vchudnov-g committed:
Rename existing batch{Set,Add}Attribute() to be more descriptive

The renaming makes it explicit they operate on string values.

vchudnov-g commented:
The "overrides" (i.e. preserving previous values n the BUILD file) were always inserting values as strings. Fixed that and added tests. PTAL.

alexander-fenster reviewed (approved):
LGTM, thank you for fixing the non-string case!
